03/tideman/tideman.c

```diff
@@ -0,0 +1,232 @@
+
+        // Query for each rank
+        for (int j = 0; j < candidate_count; j++)
+        {
+            string name = get_string("Rank %i: ", j + 1);
+
+            if (!vote(j, name, ranks))
+            {
+                printf("Invalid vote.\n");
+                return 3;
+            }
+        }
+
+        record_preferences(ranks);
+
+        printf("\n");
+    }
+
+    add_pairs();
+    sort_pairs();
+    lock_pairs();
+    print_winner();
+    return 0;
+}
+
+// Update ranks given a new vote
+bool vote(int rank, string name, int ranks[])
+{
+    for(int candidate = 0; candidate < candidate_count; candidate++) {
+
+        if(strcmp(name, candidates[candidate]) == 0) {
+            ranks[rank] = candidate;
+            return true;
+        }
+    }
+
+    return false;
+}
+
+// Update preferences given one voter's ranks
+void record_preferences(int ranks[])
+{
+    for(int candidate_a = 0; candidate_a < candidate_count; candidate_a++) {
+
+        for(int candidate_b = candidate_a + 1; candidate_b < candidate_count; candidate_b++) {
+            preferences[ranks[candidate_a]][ranks[candidate_b]]++;
+        }
+    }
+
+    return;
+}
+
+// Record pairs of candidates where one is preferred over the other
+void add_pairs(void)
+{
+    for(int candidate_a = 0; candidate_a < candidate_count; candidate_a++) {
+
+        for(int candidate_b = candidate_a + 1; candidate_b < candidate_count; candidate_b++) {
+
+            if(preferences[candidate_a][candidate_b] > preferences[candidate_b][candidate_a]) {
+
+                pairs[pair_count].winner = candidate_a;
+                pairs[pair_count].loser = candidate_b;
+                pair_count++;
+            }
+            else if(preferences[candidate_a][candidate_b] < preferences[candidate_b][candidate_a]) {
+
+                pairs[pair_count].winner = candidate_b;
+                pairs[pair_count].loser = candidate_a;
+                pair_count++;
+            }
+        }
+    }
+
+    return;
+}
+
+// Sort pairs in decreasing order by strength of victory
+void sort_pairs(void)
+{
+    for(int outerCounter = 0; outerCounter < pair_count; outerCounter++) {
+
+        for(int innerCounter = outerCounter; innerCounter < pair_count; innerCounter++) {
+
+            int strengthCurrentPair = preferences[pairs[innerCounter].winner][pairs[innerCounter].loser];
+            int strengthNextPair = preferences[pairs[innerCounter + 1].winner][pairs[innerCounter + 1].loser];
+
+            if(strengthCurrentPair < strengthNextPair) {
+
+                pair temporaryPair; temporaryPair.winner = pairs[innerCounter].winner; temporaryPair.loser = pairs[innerCounter].loser;
+
+                pairs[innerCounter].winner = pairs[innerCounter + 1].winner;
+                pairs[innerCounter].loser = pairs[innerCounter + 1].loser;
+
+                pairs[innerCounter + 1].winner = temporaryPair.winner;
+                pairs[innerCounter + 1].loser = temporaryPair.loser;
```

Comment:
```suggestion
            if(strengthCurrentPair < strengthNextPair) {
                pair temporaryPair = pairs[pair_number];
                pairs[pair_number] = pairs[pair_number+1];
                pairs[pair_number+1] = temporaryPair;
```
Du kannst das meine ich auch so machen, da somit winner und loser gleichzeitig korrekt übergeben werden